Add worker-loader

TimDaub opened the pull request:
- Fixes #814 
- Superseeds #757

Few comments
- I tried rebasing on the latest master but I got confused about the package-lock.json file. I'd appreciate some help. I'm not sure how to solve all of this properly.
- I ended up using https://github.com/surma/rollup-plugin-off-main-thread which @developit recommended in #757 
- I didn't put TypeScript support because I don't know TypeScript. I've made a note in the readme that it only works with JS builds for now. Maybe someone can add TS in a separate PR...

🤞🏼 that somehow this gets merged!

developit commented on README.md:
```diff
@@ -243,6 +261,7 @@ Options
 	--jsxImportSource  Specify the automatic import source for JSX like preact
 	--tsconfig         Specify the path to a custom tsconfig.json
 	--css-modules      Configures .css to be treated as modules (default: null)
+	--worker-loader    Generate inline worker blobs  (default false)
```
```suggestion
	--workers          Bundle module workers - see https://git.io/J3oSF  (default false)
```

developit commented on src/index.js:
```diff
@@ -366,6 +367,7 @@ function createConfig(options, entry, format, writeMeta) {
 			: () => resolve(options.cwd, 'mangle.json');
 
 	const useTypescript = extname(entry) === '.ts' || extname(entry) === '.tsx';
+	const useWorkerLoader = options['worker-loader'] !== false;
```
```suggestion
	const useWorkerLoader = options.workers !== false;
```

developit commented on src/prog.js:
```diff
@@ -56,6 +56,7 @@ export default handler => {
 		.option('--name', 'Specify name exposed in UMD builds')
 		.option('--cwd', 'Use an alternative working directory', '.')
 		.option('--sourcemap', 'Generate source map', true)
+		.option('--worker-loader', 'Generate inline worker blobs', false)
```
```suggestion
		.option('--workers', 'Bundle module workers - see https://git.io/J3oSF', false)
```

TimDaub replied:
for some reason eslint doesn't like it. I don't understand its error message:

```
59:11  error  Replace `'--workers',·'Bundle·module·workers·-·see·https://git.io/J3oSF',·false` with `⏎↹↹↹'--workers',⏎↹↹↹'Bundle·module·workers·-·see·h
ttps://git.io/J3oSF',⏎↹↹↹false,⏎↹↹`  prettier/prettier
```

developit replied:
heh - it just wants the message to be broken across multiple lines (kinda silly in this case, but alas).

TimDaub replied:
OK, makes sense.

developit commented on test/__snapshots__/index.test.js.snap:
```diff
@@ -2745,3 +2745,279 @@ exports[`fixtures build ts-module with microbundle 7`] = `
 `;
+
+exports[`fixtures build worker-loader with microbundle 1`] = `
+"Used script: microbundle -f modern,es --worker-loader
```
```suggestion
"Used script: microbundle -f modern,es --workers
```

developit commented on test/__snapshots__/index.test.js.snap:
```diff
@@ -2745,3 +2745,279 @@ exports[`fixtures build ts-module with microbundle 7`] = `
+      readme.markdown
+      test
+        all_bool.js
+        bool.js
+        dash.js
+        default_bool.js
+        dotted.js
+        kv_short.js
+        long.js
+        num.js
+        parse.js
+        parse_modified.js
+        proto.js
+        short.js
+        stop_early.js
+        unknown.js
+        whitespace.js
+    sourcemap-codec
+      CHANGELOG.md
+      LICENSE
+      README.md
+      dist
+        sourcemap-codec.es.js
+        sourcemap-codec.es.js.map
+        sourcemap-codec.umd.js
+        sourcemap-codec.umd.js.map
+        types
+          sourcemap-codec.d.ts
+      package.json
+    tippex
+      CHANGELOG.md
+      README.md
+      dist
+        tippex.es.js
+        tippex.es.js.map
+        tippex.umd.js
+        tippex.umd.js.map
+      package.json
+      src
+        index.js
```
not sure why node_modules is being included here - maybe forgot to delete it before running the fixture?

TimDaub replied:
oh, that could be. I'll take a look

TimDaub replied:
Should be resolved in f155eee

developit commented on test/fixtures/worker-loader/package.json:
```diff
@@ -0,0 +1,6 @@
+{
+  "name": "worker-loader",
+  "scripts": {
+    "build": "microbundle -f modern,es --worker-loader"
```
```suggestion
    "build": "microbundle -f modern,es --workers"
```

developit commented on README.md:
````diff
@@ -192,6 +192,24 @@ By default Microbundle outputs multiple bundles, one bundle per format. A single
 microbundle -i lib/main.js -o dist/bundle.js --no-pkg-main -f umd
 ```
 
+### Building web workers as inline blobs
+
+Microbundle is able to detect and bundle workers when building with format `es`
+or `modern`. Consider the following example:
+
+```js
+const worker = new Worker('./worker.js', { type: 'module' });
+```
+
+And build it like this:
+
+```bash
+microbundle --worker-loader
+```
````
````suggestion
### Building Module Workers

Microbundle is able to detect and bundle Module Workers when generating bundles in the
`es`, `umd` and `modern` formats. To use this feature, instantiate your Web Worker as follows:

```js
worker = new Worker(new URL('./worker.js', import.meta.url), { type: 'module' });
// or simply:
worker = new Worker('./worker.js', { type: 'module' });
```

... then add the `--workers` flag to your build command:

```bash
microbundle --workers
```
````

TimDaub commented:
@developit I addressed your comments. I'm not sure I'm capable of rebasing though. Last time I tried sketchy things happened in package-lock.json that I wasn't able to address. What now?

ForsakenHarmony commented:
> Last time I tried sketchy things happened in package-lock.json that I wasn't able to address.

You generally have to accept the remote one and then after create a new commit / amend a commit with a file you generated after by running `npm install`

developit commented:
I'll try to get this rebased this week (I'm also terrible at it!).

This will need to go into a new breaking release (minor since microbundle is still pre-1.0), but I think that should be fine.

developit commented:
Just wanted to mention I haven't forgotten about this PR, I just have been side-tracked rewriting Microbundle's filename inference.

developit commented:
This was landed in #867! 🎉
